x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2857 @@
+
+For example: in a `required_fields` update, if the `type` is updated from `unknown` in the base version to `keyword` in the `target` version, but in the same object the user has updated the base `name` of the field, the change of type will have a high probability of not making sense anymore - there is no way of knowing if the change proposed by Elastic will also apply to the field name that the user is now referring through its customization.
+
+Therefore, we should prefer to keep the user's version for the whole block.(See this example in the first element of the array in the first row of the table below).
+
+Examples: `related_integrations,` `required_fields`
+
+###### Proposed algorithm
+
+- Do element by element comparison:
+  - If element index exists in the three versions and the values:
+    - do not differ between all versions (base, current and target), **keep current** version. (A A A)
+    - do not differ between base and current, but updates target, **keep target** version. (A A B)
+    - differ between base and current, but doesn't update target, **keep current** version. (A B A)
+    - differ between base and current, updates target, **keep current** version. (A B B)
+    - differ between all version (base, current and target), **keep current** version. (A B C)
+  - If element index exists in base and current version, but not in target:
+    - If base differs from current: **keep current** version. (A B _)
+    - If base equals current: **remove element**.  (A A _)
+  - If element index exists in base and target version, but not in current:
+    - **Remove element**. (A _ B) and (A _ A)
+  - If element index exists in current and target version, but not in base:
+    - If base equals current: **add element/keep target**. (_ A A)
+    - If base differs from current: **keep both current and target**. (_ A B)
+  - If element index exists only in base version:
+    - **Remove element** (A _ _)
+  - If element index exists only in current version:
+    - **Keep current version** (_ A _)
+  - If element index exists only in target version:
+    - **Keep target version** (_ _ A)
+    
+<table>
+    <thead>
+      <tr>
+        <th style="border-right:3px solid black">Use case</th>
+        <th>Base version</th>
+        <th>Current version</th>
+        <th style="border-right:3px solid black">Target version</th>
+        <th>Merged version (output)</th>
+      </tr>
+    </thead>
+  <tbody valign="top">
+    <tr>
+      <td rowspan=2 style="border-right:3px solid black">Keep user changes if element conflicts, but apply changes in elements with clean updates <br><br> <ul><li><b>1st elem: (A B C - keep current) </li><li>2nd elem: (A A B - keep target)</b></li></ul><br><br> Same for: <br><ul><li><b>1st elem: (A B C) + 2nd elem: (A B A - keep current)</b></li><li><b>1st elem: (A B C) + 2nd elem: (A B B - keep current/target)</b></li></ul></td>
+      <td>
+      <pre>
+[
+  logs-*,
+  logstash-*
+]
```

Comment:
Why do we describe examples of changes to index patterns in the "Array of objects fields" section and not in the "Array of strings fields"? Please move it to "Array of strings fields" to a separate table.

Reply:
Can't agree with the proposed algorithm for index patterns either:

<img width="1394" alt="Screenshot 2024-03-18 at 15 12 32" src="https://github.com/elastic/kibana/assets/7359339/664ac763-91d9-4769-ad77-2a08639e8724">

In the example above I'd expect the result to be:

```ts
['cluster_one:logs-*', 'cluster_one:logstash-*', 'new-index1-*', 'new-index2-*']
```

The logic would be simple: https://github.com/elastic/kibana/pull/171856#discussion_r1528519565